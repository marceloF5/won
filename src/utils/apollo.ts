import { ApolloClient, HttpLink, NormalizedCacheObject } from '@apollo/client'
import { useMemo } from 'react'
import apolloCache from './apolloCache'
import { setContext } from '@apollo/client/link/context'
import { Session } from 'next-auth/client'

let apolloClient: ApolloClient<NormalizedCacheObject | null>

function createApolloClient(session?: Session | null) {
    const httpLink = new HttpLink({
        uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`
    })

    const authLink = setContext((_, { headers }) => {
        const authorization = session?.jwt ? `Bearer ${session.jwt}` : ''

        return { headers: { ...headers, authorization } }
    })

    return new ApolloClient({
        ssrMode: typeof window === 'undefined', // true
        link: authLink.concat(httpLink),
        cache: apolloCache
    })
}

export function initializeApollo(
    initialState = null,
    session?: Session | null
) {
    // verify if there is an apolloCliente instance. If there is no instance, it will create a new one.
    const apolloClientGlobal = apolloClient ?? createApolloClient(session)

    // verify if there is data in cache
    if (initialState) {
        apolloClientGlobal.cache.restore(initialState)
    }

    // it's always run in SSR if new cache
    if (typeof window === 'undefined') return apolloClientGlobal

    apolloClient = apolloClient ?? apolloClientGlobal

    return apolloClient
}

export function useApollo(initialState = null, session?: Session | null) {
    const store = useMemo(() => initializeApollo(initialState, session), [
        initialState,
        session
    ])

    return store
}
