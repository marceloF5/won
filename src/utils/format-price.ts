export default function formatPrice(price: number | bigint): string {
    if (price === 0) return 'FREE'

    return new Intl.NumberFormat('en', {
        style: 'currency',
        currency: 'USD'
    }).format(price)
}
