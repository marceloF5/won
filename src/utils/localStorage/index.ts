const APP_KEY = 'WONGAMES'

export function getStorageItem(key: string) {
    // there is no window via Server/Static
    if (typeof window === 'undefined') return

    const data = window.localStorage.getItem(`${APP_KEY}_${key}`)

    return JSON.parse(data!)
}

export function setStorageItem(key: string, value: string[]) {
    // there is no window via Server/Static
    if (typeof window === 'undefined') return

    const data = JSON.stringify(value)

    return window.localStorage.setItem(`${APP_KEY}_${key}`, data)
}
