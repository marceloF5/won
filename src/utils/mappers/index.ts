import {
    QueryHome_banners,
    QueryHome_sections_freeGames_highlight
} from 'graphql/generated/QueryHome'
import { QueryGames_games } from 'graphql/generated/QueryGames'
import formatPrice from '../format-price'
import { QueryWishlist_wishlists_games } from 'graphql/generated/QueryWishlist'

export const bannerMapper = (banners: QueryHome_banners[]) => {
    return banners.map((banner) => ({
        img: `http://localhost:1337${banner.image!.url}`,
        title: banner.title,
        subtitle: banner.subtitle,
        buttonLabel: banner.button?.label,
        buttonLink: banner.button?.link,
        ...(banner.ribbon && {
            ribbon: banner.ribbon.text,
            ribbonColor: banner.ribbon.color,
            ribbonSize: banner.ribbon.size
        })
    }))
}

export const gamesMapper = (
    games:
        | QueryGames_games[]
        | QueryWishlist_wishlists_games[]
        | null
        | undefined
) => {
    return games
        ? games.map((game) => ({
              id: game.id,
              slug: game.slug,
              title: game.name,
              developer: game.developers[0].name,
              img: `http://localhost:1337${game.cover?.url}`,
              price: game.price
          }))
        : []
}

export const highLightMapper = (
    highlight: QueryHome_sections_freeGames_highlight | null | undefined
) => {
    return (
        highlight && {
            title: highlight.title,
            subtitle: highlight.subtitle,
            backgroundImage: `http://localhost:1337${highlight.background?.url}`,
            floatImage: `http://localhost:1337${highlight.floatImage?.url}`,
            buttonLabel: highlight.buttonLabel,
            buttonLink: highlight.buttonLink,
            alignment: highlight.alignment
        }
    )
}

export const cartMapper = (games: QueryGames_games[] | undefined) => {
    return games
        ? games.map((game) => ({
              id: game.id,
              title: game.name,
              img: `http://localhost:1337${game.cover?.url}`,
              price: formatPrice(game.price)
          }))
        : []
}
