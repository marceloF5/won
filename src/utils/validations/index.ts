import Joi from 'joi'
import { UsersPermissionsRegisterInput } from 'graphql/generated/globalTypes'

type ResetValidateParams = { confirm_password: string } & Pick<
    UsersPermissionsRegisterInput,
    'password'
>

export type FieldErrors = {
    [key: string]: string
}

const fieldsValidations = {
    username: Joi.string().min(5).required(),
    email: Joi.string()
        .email({ tlds: { allow: false } })
        .required(),
    password: Joi.string().required(),
    confirm_password: Joi.string()
        .valid(Joi.ref('password'))
        .required()
        .messages({
            'any.only': 'Confirm password does not match with password'
        })
}

function getFieldErrors(objErrors: Joi.ValidationResult) {
    const errors: FieldErrors = {}

    if (objErrors.error) {
        objErrors.error.details.forEach((err) => {
            errors[err.path.join('.')] = err.message
        })
    }

    return errors
}

export function signInValidate(
    values: Omit<UsersPermissionsRegisterInput, 'username'>
) {
    const { email, password } = fieldsValidations
    const schema = Joi.object({ email, password })

    return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

export function signUpValidate(values: UsersPermissionsRegisterInput) {
    const schema = Joi.object(fieldsValidations)

    return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

export function forgotValidate(
    values: Pick<UsersPermissionsRegisterInput, 'email'>
) {
    const { email } = fieldsValidations
    const schema = Joi.object({ email })

    return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

export function resetValidate(values: ResetValidateParams) {
    const { password, confirm_password } = fieldsValidations
    const schema = Joi.object({ password, confirm_password })

    return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
