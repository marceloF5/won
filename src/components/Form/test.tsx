import { render } from 'utils/test-utils'

import { FormWrapper, FormLink } from '.'

describe('<Form />', () => {
    it('should render the heading', () => {
        const { container } = render(
            <FormWrapper>
                <FormLink>My link</FormLink>
            </FormWrapper>
        )

        expect(container.parentElement).toMatchSnapshot()
    })
})
