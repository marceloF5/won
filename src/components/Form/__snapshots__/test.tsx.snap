// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render the heading 1`] = `
.c0 .sc-AxheI {
  margin: 3.2rem auto 1.6rem;
}

.c0 .sc-AxjAm {
  margin: 0.8rem 0;
}

.c1 {
  color: #030517;
  font-size: 1.4rem;
  text-align: center;
}

.c1 a {
  border-bottom: 0.1rem solid #3cd3c1;
  color: #3cd3c1;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color,border,0.1s ease-in-out;
  transition: color,border,0.1s ease-in-out;
}

.c1 a:hover {
  color: #29b3a3;
  border-bottom: 0.1rem solid #29b3a3;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
      >
        My link
      </div>
    </div>
  </div>
</body>
`
